File: fs/jagex/DiskStorage.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

/*
 * Copyright (c) 2016-2017, Adam <Adam@sigterm.info>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
namespace OSRSCache.fs.jagex
{
	using Ints = com.google.common.primitives.Ints;
	using Archive = OSRSCache.fs.Archive;
	using Container = OSRSCache.fs.Container;
	using Index = OSRSCache.fs.Index;
	using Storage = OSRSCache.fs.Storage;
	using Store = OSRSCache.fs.Store;
	using ArchiveData = OSRSCache.index.ArchiveData;
	using IndexData = OSRSCache.index.IndexData;
	using Crc32 = OSRSCache.util.Crc32;


	public class DiskStorage : Storage
	{
		private const string MAIN_FILE_CACHE_DAT = "main_file_cache.dat2";
		private const string MAIN_FILE_CACHE_IDX = "main_file_cache.idx";

		private readonly string folder;

		private readonly DataFile data;
		private readonly IndexFile index255;
		private readonly IList<IndexFile> indexFiles = new List<IndexFile>();

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: public DiskStorage(java.io.File folder) throws java.io.IOException
		public DiskStorage(string folder)
		{
			this.folder = folder;

			this.data = new DataFile($"{folder}/{MAIN_FILE_CACHE_DAT}");
			this.index255 = new IndexFile(255, $"{folder}/{MAIN_FILE_CACHE_IDX}255");
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: @Override public void init(OSRSCache.fs.Store store) throws java.io.IOException
		public virtual void init(Store store)
		{
			for (int i = 0; i < index255.IndexCount; ++i)
			{
				store.addIndex(i);
				getIndex(i);
			}

			Debug.Assert(store.Indexes.Count == indexFiles.Count);
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: @Override public void close() throws java.io.IOException
		public virtual void close()
		{
			data.Dispose();
			index255.Dispose();
			foreach (IndexFile indexFile in indexFiles)
			{
				indexFile.Dispose();
			}
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: private IndexFile getIndex(int i) throws java.io.FileNotFoundException
		private IndexFile getIndex(int i)
		{
			foreach (IndexFile indexFile in indexFiles)
			{
				if (indexFile.IndexFileId == i)
				{
					return indexFile;
				}
			}

			IndexFile _indexFile = new IndexFile(i, $"{folder}/{MAIN_FILE_CACHE_IDX}{i}"); 
			indexFiles.Add(_indexFile);
			return _indexFile;
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: @Override public void load(OSRSCache.fs.Store store) throws java.io.IOException
		public virtual void load(Store store)
		{
			foreach (Index index in store.Indexes)
			{
				loadIndex(index);
			}
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: public byte[] readIndex(int indexId) throws java.io.IOException
		public virtual sbyte[] readIndex(int indexId)
		{
			IndexEntry entry = index255.read(indexId);
			if (entry == null)
			{
				return null;
			}

			sbyte[] indexData = data.read(index255.IndexFileId, entry.Id, entry.Sector, entry.Length);
			return indexData;
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: private void loadIndex(OSRSCache.fs.Index index) throws java.io.IOException
		private void loadIndex(Index index)
		{
			Console.WriteLine("Loading index {}", index.Id);

			sbyte[] indexData = readIndex(index.Id);
			if (indexData == null)
			{
				return;
			}

			Container res = Container.decompress(indexData, null);
			sbyte[] data = res.data;

			IndexData id = new IndexData();
			id.load(data);

			index.Protocol = id.Protocol;
			index.Revision = id.Revision;
			index.Named = id.Named;

			foreach (ArchiveData ad in id.Archives)
			{
				Archive archive = index.addArchive(ad.Id);
				archive.NameHash = ad.NameHash;
				archive.Crc = ad.Crc;
				archive.Revision = ad.Revision;
				archive.FileData = ad.Files;

				Debug.Assert(ad.Files.Length > 0);
			}

			index.Crc = res.crc;
			index.Compression = res.compression;
			Debug.Assert(res.revision == -1);
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: @Override public byte[] loadArchive(OSRSCache.fs.Archive archive) throws java.io.IOException
		public virtual sbyte[] loadArchive(Archive archive)
		{
			Index index = archive.Index;
			IndexFile indexFile = getIndex(index.Id);

			Debug.Assert(indexFile.IndexFileId == index.Id);

			IndexEntry entry = indexFile.read(archive.ArchiveId);
			if (entry == null)
			{
				Console.WriteLine("can't read archive " + archive.ArchiveId + " from index " + index.Id);
				return null;
			}

			Debug.Assert(entry.Id == archive.ArchiveId);

			Console.WriteLine("Loading archive {} for index {} from sector {} length {}", archive.ArchiveId, index.Id, entry.Sector, entry.Length);

			sbyte[] archiveData = data.read(index.Id, entry.Id, entry.Sector, entry.Length);
			return archiveData;
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: @Override public void save(OSRSCache.fs.Store store) throws java.io.IOException
		public virtual void save(Store store)
		{
			Console.WriteLine("Saving store");

			foreach (Index i in store.Indexes)
			{
				saveIndex(i);
			}
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: private void saveIndex(OSRSCache.fs.Index index) throws java.io.IOException
		private void saveIndex(Index index)
		{
			IndexData indexData = index.toIndexData();
			sbyte[] data = indexData.writeIndexData();

			Container container = new Container(index.Compression, -1); // index data revision is always -1
			container.compress(data, null);
			sbyte[] compressedData = container.data;
			DataFileWriteResult res = this.data.write(index255.IndexFileId, index.Id, compressedData);

			index255.write(new IndexEntry(index255, index.Id, res.sector, res.compressedLength));

			Crc32 crc = new Crc32();
			crc.update(compressedData, 0, compressedData.Length);
			index.Crc = crc.Hash;
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: @Override public void saveArchive(OSRSCache.fs.Archive a, byte[] archiveData) throws java.io.IOException
		public virtual void saveArchive(Archive a, sbyte[] archiveData)
		{
			Index index = a.Index;
			IndexFile indexFile = getIndex(index.Id);
			Debug.Assert(indexFile.IndexFileId == index.Id);

			DataFileWriteResult res = data.write(index.Id, a.ArchiveId, archiveData);
			indexFile.write(new IndexEntry(indexFile, a.ArchiveId, res.sector, res.compressedLength));

			sbyte compression = archiveData[0];
			int compressedSize = Ints.fromBytes(archiveData[1], archiveData[2], archiveData[3], archiveData[4]);

			// don't crc the appended revision, if it is there
			int length = 1 + 4 + compressedSize + (compression != CompressionType.NONE ? 4 : 0);

			Crc32 crc = new Crc32();
			crc.update(archiveData, 0, length);
			a.Crc = crc.Hash;

			Console.WriteLine("Saved archive {}/{} at sector {}, compressed length {}", index.Id, a.ArchiveId, res.sector, res.compressedLength);
		}
	}

}

File: src/main/java/net/runelite/cache/fs/jagex/DiskStorage.cpp
#include "DiskStorage.h"
#include "IndexEntry.h"
#include "DataFileWriteResult.h"
#include "CompressionType.h"

namespace net::runelite::cache::fs::jagex
{
	using Ints = com::google::common::primitives::Ints;
	using File = java::io::File;
	using FileNotFoundException = java::io::FileNotFoundException;
	using IOException = java::io::IOException;
	using ArrayList = java::util::ArrayList;
	using List = java::util::List;
	using Archive = net::runelite::cache::fs::Archive;
	using Container = net::runelite::cache::fs::Container;
	using Index = net::runelite::cache::fs::Index;
	using Storage = net::runelite::cache::fs::Storage;
	using Store = net::runelite::cache::fs::Store;
	using ArchiveData = net::runelite::cache::index::ArchiveData;
	using IndexData = net::runelite::cache::index::IndexData;
	using Crc32 = net::runelite::cache::util::Crc32;
	using Logger = org::slf4j::Logger;
	using LoggerFactory = org::slf4j::LoggerFactory;
const std::shared_ptr<org::slf4j::Logger> DiskStorage::logger = org::slf4j::LoggerFactory::getLogger(DiskStorage::class);
const std::wstring DiskStorage::MAIN_FILE_CACHE_DAT = L"main_file_cache.dat2";
const std::wstring DiskStorage::MAIN_FILE_CACHE_IDX = L"main_file_cache.idx";

	DiskStorage::DiskStorage(std::shared_ptr<File> folder) : folder(folder), data(std::make_shared<DataFile>(std::make_shared<File>(folder, MAIN_FILE_CACHE_DAT))), index255(std::make_shared<IndexFile>(255, std::make_shared<File>(folder, MAIN_FILE_CACHE_IDX + L"255")))
	{

	}

	void DiskStorage::init(std::shared_ptr<Store> store)
	{
		for (int i = 0; i < index255->getIndexCount(); ++i)
		{
			store->addIndex(i);
			getIndex(i);
		}

		assert(store->getIndexes().size() == indexFiles.size());
	}

	void DiskStorage::close()
	{
		data->close();
		index255->close();
		for (auto indexFile : indexFiles)
		{
			indexFile->close();
		}
	}

	std::shared_ptr<IndexFile> DiskStorage::getIndex(int i)
	{
		for (auto indexFile : indexFiles)
		{
			if (indexFile->getIndexFileId() == i)
			{
				return indexFile;
			}
		}

		std::shared_ptr<IndexFile> indexFile = std::make_shared<IndexFile>(i, std::make_shared<File>(folder, MAIN_FILE_CACHE_IDX + std::to_wstring(i)));
		indexFiles.push_back(indexFile);
		return indexFile;
	}

	void DiskStorage::load(std::shared_ptr<Store> store)
	{
		for (auto index : store->getIndexes())
		{
			loadIndex(index);
		}
	}

	std::vector<signed char> DiskStorage::readIndex(int indexId)
	{
		std::shared_ptr<IndexEntry> entry = index255->read(indexId);
		if (entry == nullptr)
		{
			return std::vector<signed char>();
		}

		std::vector<signed char> indexData = data->read(index255->getIndexFileId(), entry->getId(), entry->getSector(), entry->getLength());
		return indexData;
	}

	void DiskStorage::loadIndex(std::shared_ptr<Index> index)
	{
		logger->trace(L"Loading index {}", index->getId());

		std::vector<signed char> indexData = readIndex(index->getId());
		if (indexData.empty())
		{
			return;
		}

		std::shared_ptr<Container> res = Container::decompress(indexData, std::vector<int>());
		std::vector<signed char> data = res->data;

		std::shared_ptr<IndexData> id = std::make_shared<IndexData>();
		id->load(data);

		index->setProtocol(id->getProtocol());
		index->setRevision(id->getRevision());
		index->setNamed(id->isNamed());

		for (auto ad : id->getArchives())
		{
			std::shared_ptr<Archive> archive = index->addArchive(ad->getId());
			archive->setNameHash(ad->getNameHash());
			archive->setCrc(ad->getCrc());
			archive->setRevision(ad->getRevision());
			archive->setFileData(ad->getFiles());

			assert(ad->getFiles().size() > 0);
		}

		index->setCrc(res->crc);
		index->setCompression(res->compression);
		assert(res->revision == -1);
	}

	std::vector<signed char> DiskStorage::loadArchive(std::shared_ptr<Archive> archive)
	{
		std::shared_ptr<Index> index = archive->getIndex();
		std::shared_ptr<IndexFile> indexFile = getIndex(index->getId());

		assert(indexFile->getIndexFileId() == index->getId());

		std::shared_ptr<IndexEntry> entry = indexFile->read(archive->getArchiveId());
		if (entry == nullptr)
		{
			logger->debug(L"can't read archive " + std::to_wstring(archive->getArchiveId()) + L" from index " + std::to_wstring(index->getId()));
			return std::vector<signed char>();
		}

		assert(entry->getId() == archive->getArchiveId());

		logger->trace(L"Loading archive {} for index {} from sector {} length {}", archive->getArchiveId(), index->getId(), entry->getSector(), entry->getLength());

		std::vector<signed char> archiveData = data->read(index->getId(), entry->getId(), entry->getSector(), entry->getLength());
		return archiveData;
	}

	void DiskStorage::save(std::shared_ptr<Store> store)
	{
		logger->debug(L"Saving store");

		for (auto i : store->getIndexes())
		{
			saveIndex(i);
		}
	}

	void DiskStorage::saveIndex(std::shared_ptr<Index> index)
	{
		std::shared_ptr<IndexData> indexData = index->toIndexData();
		std::vector<signed char> data = indexData->writeIndexData();

		std::shared_ptr<Container> container = std::make_shared<Container>(index->getCompression(), -1); // index data revision is always -1
		container->compress(data, std::vector<int>());
		std::vector<signed char> compressedData = container->data;
		std::shared_ptr<DataFileWriteResult> res = this->data->write(index255->getIndexFileId(), index->getId(), compressedData);

		index255->write(std::make_shared<IndexEntry>(index255, index->getId(), res->sector, res->compressedLength));

		std::shared_ptr<Crc32> crc = std::make_shared<Crc32>();
		crc->update(compressedData, 0, compressedData.size());
		index->setCrc(crc->getHash());
	}

	void DiskStorage::saveArchive(std::shared_ptr<Archive> a, std::vector<signed char> &archiveData)
	{
		std::shared_ptr<Index> index = a->getIndex();
		std::shared_ptr<IndexFile> indexFile = getIndex(index->getId());
		assert(indexFile->getIndexFileId() == index->getId());

		std::shared_ptr<DataFileWriteResult> res = data->write(index->getId(), a->getArchiveId(), archiveData);
		indexFile->write(std::make_shared<IndexEntry>(indexFile, a->getArchiveId(), res->sector, res->compressedLength));

		signed char compression = archiveData[0];
		int compressedSize = Ints::fromBytes(archiveData[1], archiveData[2], archiveData[3], archiveData[4]);

		// don't crc the appended revision, if it is there
		int length = 1 + 4 + compressedSize + (compression != CompressionType::NONE ? 4 : 0);

		std::shared_ptr<Crc32> crc = std::make_shared<Crc32>();
		crc->update(archiveData, 0, length);
		a->setCrc(crc->getHash());

		logger->trace(L"Saved archive {}/{} at sector {}, compressed length {}", index->getId(), a->getArchiveId(), res->sector, res->compressedLength);
	}
}
